net/retriever/retriever.go

```diff
@@ -58,14 +59,14 @@ type offlineMessage struct {
 	env  pb.Envelope
 }
 
-func NewMessageRetriever(db repo.Datastore, ctx commands.Context, node *core.IpfsNode, bm *net.BanManager, service net.NetworkService, prefixLen int, pushNodes []peer.ID, dialer proxy.Dialer, sendAck func(peerId string, pointerID peer.ID) error) *MessageRetriever {
+func NewMessageRetriever(db repo.Datastore, ctx commands.Context, node *core.IpfsNode, bm *net.BanManager, service net.NetworkService, prefixLen int, pushNodes []peer.ID, dialer proxy.Dialer, sendAck func(peerId string, pointerID peer.ID) error, sendError func(peerId string, k *libp2p.PubKey, errorMessage pb.Message) error) *MessageRetriever {
```

Comment:
Option struct as an argument to this function please? This function signature is out of control. If you can't write a function call without using intellisense, there's probably cognitive overhead that can be reduced. The reason I like struct options is because they are self-documenting and (with good field names) make the arguments easier to understand without referring to how they were made.